lib/results.js

```diff
@@ -0,0 +1,157 @@
+ * harness status.
+ *
+ * [0]: https://web-platform-tests.org/writing-tests/testharness-api.html#single-page-tests
+ */
+
+const Git = require('nodegit');
+
+// Map from object id to { "trees": { ... }, "tests": { ... } } objects.
+const treeCache = {};
+
+// Map from object id to { "status": "OK", ... } objects.
+const testCache = {};
+
+// Convert a git object id to a key in the above maps.
+function oidToKey(oid) {
+  return oid.tostrS();
+}
+
+async function readResults(entry) {
+  if (!entry.isBlob()) {
+    throw new TypeError('y no Blob?');
+  }
+
+  const key = oidToKey(entry.id());
+
+  const cachedTest = testCache[key];
+  if (cachedTest) {
+    return cachedTest;
+  }
+
+  const blob = await entry.getBlob();
+  const buffer = blob.content();
+  const results = JSON.parse(buffer);
+
+  testCache[key] = results;
+  return results;
+}
+
+async function readTree(treeOrEntry) {
+  let tree, entry, oid;
+  if (treeOrEntry instanceof Git.Tree) {
+    tree = treeOrEntry;
+    oid = tree.id();
+  } else {
+    if (!(treeOrEntry instanceof Git.TreeEntry) || !treeOrEntry.isTree()) {
+      throw new TypeError('y no Tree or TreeEntry?');
+    }
+    entry = treeOrEntry;
+    oid = entry.id();
+  }
+
+  const key = oidToKey(oid);
+
+  const cachedTree = treeCache[key];
+  if (cachedTree) {
+    return cachedTree;
+  }
+
+  const newTree = {
+    trees: {},
+    tests: {},
+  };
+
+  if (!tree) {
+    tree = await entry.getTree();
+  }
+
+  for (const entry of tree.entries()) {
+    if (entry.isTree()) {
+      newTree.trees[entry.name()] = await readTree(entry);
+    } else if (entry.isBlob()) {
+      let name = entry.name();
+      if (!name.endsWith('.json')) {
+        throw new Error('y not .json?');
+      }
+      name = decodeURIComponent(name.substr(0, name.length - 5));
+      newTree.tests[name] = await readResults(entry);
+    } else {
+      throw new TypeError('y not tree or blob?')
+    }
+  }
+
+  treeCache[key] = newTree;
+  return newTree;
+}
+
+// Read the tree for a given run fully into memory, converting it into our
+// internal representation (see module documentation).
+async function getGitTree(repo, run) {
+  const commit = await repo.getReferenceCommit(`refs/tags/run/${run.id}/results`);
+  const tree = await commit.getTree();
+
+  return readTree(tree);
+}
+
+// Return a set of run ids, determined from the tags of the git repo.
+async function getLocalRunIds(repo) {
+  const refs = await repo.getReferences();
+  const tags = refs.filter(ref => ref.isTag());
+  tags.sort();
+
+  return new Set(tags.map(tag => {
+    // format is refs/tags/run/6286849043595264/results
+    const parts = tag.toString().split('/');
+    return Number(parts[3]);
+  }));
+}
+
+// Walks an input tree in depth-first order, calling the visitor function on
+// each test in the tree. The visitor function should be of the form:
+//   visitor(path, test_name, test_results)
+//
+// Where test_results is a json dictionary as described in the module
```

Comment:
```suggestion
// Where test_results is an object as described in the module
```

JSON is a string, the actual value is an object :)